### DISCUSSION TITLE
How to migrate SQL-Ledger 3.2.12 to LedgerSMB on Ubuntu

### DISCUSSION THREAD

**Discussion by ondohotola** [ledgersmb/LedgerSMB#7435] [Q&A]
Hi,

How would one go about migrating a number of sets of books, some of which have up to 20 years of data from the latest SQL-Ledger version to LedgerSMB?

greetings, el

**Comment by ehuelsmann** [comment]
Hi,

Thank you for contacting the project. Originally, migration was dead simple ( around SQL Ledger 2.8 and LedgerSMB 1.1 and 1.2 ). These days, migration is a bit more involved, but still possible and definitely something the project likes to support. Even more: as I write this, there's an SL3.2 related improvement request open (see https://github.com/ledgersmb/LedgerSMB/pull/7422).

We're happy to help, either through this discussion forum and/or on the Matrix chat channel (https://app.element.io/#/room/#ledgersmb:matrix.org) for a more interactive experience.

Can you tell a bit more about the environment you're using? That is:

* Are you using Windows/Linux/FreeBSD? 
* If Linux, which distribution (and which version)?
* Which version of PostgreSQL are you currently using?
* To which version of LedgerSMB are you planning to migrate? (If I can suggest something: choose 1.10)
* One of the simplest installation means could be to install LedgerSMB using Docker (depending on what your experience level with it is); do you have Docker experience?

Regards,

Erik.

**Reply by ehuelsmann** [reply]
Sorry: "On Ubuntu"...

Missed that part. Still: the other questions stand. (Also, which version of Ubuntu?)

**Reply by ehuelsmann** [reply]
There's another way, by the way, which is to allow me access to your environment and do the migration for you. (I'm assuming fixes/additions to the migration code are required for your data-set.) Or to provide me the data, if you would rather not provide access to your environment, and migrate it for you, after which I hand the data back. The required enhancements to the migration code will end up in the public repository either way.

**Reply by ylavoie** [reply]
I do have successfully migrated SL3.2, code will be added to LedgerSMB by the end of the week. It currently does everything but Payroll.

**Comment by ehuelsmann** [comment]
@ondohotola if you have a data set of similar quality as @ylavoie, these would be the steps (to be executed on a copy of the database):

1. Install LedgerSMB (or run a Docker image)
2. Navigate to setup.pl (usually http://localhost/setup.pl) and log in with `postgres` (make sure to set a password for postgres-logging-in-over-IP), provide the name of the database you want to migrate in the "database" field.
3. In the resulting screen it will say "SQL Ledger 3.2 database identified" (or something alike), followed by a button which says "Upgrade". Clicking that button will start the upgrade procedure.

It's definitely possible that data quality issues are identified during the upgrade process. In case we already ran into issues like it before, the upgrade will ask you which direction you want to go repairing the data. If not, you'll likely need to contact us either here, on the mailing lists or on the chat.

Regards,

Erik.

**Comment by ondohotola** [comment]
Thanks.

I run Ubuntu 20.04 LTS with PostgreSQL 14 on one and 12 on the others and 22.04 with PostgreSQL 14. I am fairly computer literate, can install Ubuntu and so on. I will not use Docker and will migrate to whatever version the consensus here says makes most sense.

I would love to see the 3.2 migration script from @ylavoie.

In case you are not aware, Mr Simader's widow has informed me this week by email of his passing, so I am sure more users will look at this particular migration, and it will be helpful to have such a script.

**Reply by ehuelsmann** [reply]
> In case you are not aware, Mr Simader's widow has informed me this week by email of his passing, so I am sure more users will look at this particular migration, and it will be helpful to have such a script.

Thanks for notifying us. I became aware last week. It didn't feel appropriate to post on the SQL Ledger forum about the existence of our project, especially since the two parted in anger (although I wasn't there at the time and it's 17 years ago...).

But we'll gladly help anybody who wants to migrate, either for this reason or any other reason, really.

**Reply by ehuelsmann** [reply]
@ondohotola, just checked with @ylavoie. He used the procedure outlined above. He confirmed he did not use any additional migration steps.

Concluding: We should be helping you getting installed with LedgerSMB so you can try the migration.

**Reply by ehuelsmann** [reply]
One note he just sent me through private chat: Vouchers and Payroll are currently not included in the migration; if you need those, we need to develop that.

**Comment by ondohotola** [comment]
Erik

thank you.

I have not used these and do not foresee their use.

I have rewritten my invoices to use the KOMAScript (scrartcl) and
made a number of modifications. Should the templating have
changed I can easily rewrite these.


What I do use a lot is "Import Sales Invoice" and lesser so
"Import Payments", where I generate the CSV by way of shell
scripts.  Does LedgerSMB Support some or this form of importing?
I can easily rewrite my scripts.


I also noted some references to OFX for bank reconciliation. As
one of my banks does not support OFX I have become quite familiar
with the Python package CSV2OFX which I use as an intermediary
step in generating the API scripts. Direct reading of OFX would
reduce the complexity significantly.


If any or all of this is covered int he documentation, please feel free
to point me to it, I am not averse to RTFM :-)-O


I will in any case update the one server using PostgreSQL 12 to
14 before migrating so I use the same everywhere.

The servers will be upgraded in the course of the year to Ubuntu
22.04 in the course of the year, so eventually all will be
streamlined.

The first migration will be on recently set up books with very
few transactions on 14. I'll attempt the installation of LedgerSMB
there first, and would appreciate a pointer to where this is
described. Should

Should some or all of this be more suitable for off line
discussion, feel free to contact me on my email address
directly, though I am happy to continue here. Since that 
will be obfuscated by github you can use  'el' [at] 'lisse' '.na'

greetings, el


On 01/06/2023 21:59, Erik Huelsmann wrote:
> One note he just sent me through private chat: Vouchers and
> Payroll are currently not included in the migration; if you need
> those, we need to develop that.
[...]
